File: api_nestjs_distributed_order_platform/src/application/use-case/order/confirm-order.use-case.ts
import { BadRequestException, Inject, Injectable, Logger, NotFoundException } from "@nestjs/common";
import type { EventPublisherRepositoryInterface } from "src/application/port/event-publisher.repository";
import type { OrderRepositoryInterface } from "src/application/port/order.repository";
import { OrderEntity } from "src/domain/entities/order.entity";
import { ORDER_REPOSITORY } from "src/shared/tokens_nest/order.token";
import { EVENT_PUBLISHER } from "src/shared/tokens_nest/rabbitmq.token";

@Injectable()
export class ConfirmOrderUseCase {
    private readonly logger = new Logger(ConfirmOrderUseCase.name);

    constructor(
        @Inject(ORDER_REPOSITORY)
        private readonly orderRepository: OrderRepositoryInterface,
        @Inject(EVENT_PUBLISHER)
        private readonly eventPublisher: EventPublisherRepositoryInterface,
    ) {}

    async execute(orderId: string): Promise<OrderEntity> {
        const order = await this.orderRepository.findById(orderId);
        if (!order) {
            throw new NotFoundException('Order not found');
        }

        try {
            order.confirmOrder();
        } catch (error) {
            throw new BadRequestException(
                error instanceof Error ? error.message : 'Unable to confirm order',
            );
        }

        await this.orderRepository.update(order);

        try{
            await this.eventPublisher.publishConfirmedOrder(order);
        }catch(error){
            this.logger.error(
                `Failed to publish ConfirmedOrder for order ${order.getId()}`,
                error instanceof Error ? error.stack : String(error),
            );
        }
        
        return order;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/order/ConfirmOrderUseCase.java
package com.project.order.application.usecase.order;

import com.project.order.application.exception.BusinessRuleException;
import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.EventPublisherRepository;
import com.project.order.application.port.OrderRepository;
import com.project.order.domain.model.Order;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ConfirmOrderUseCase {

    private static final Logger log = LoggerFactory.getLogger(ConfirmOrderUseCase.class);

    private final OrderRepository orderRepository;
    private final EventPublisherRepository eventPublisher;

    public ConfirmOrderUseCase(OrderRepository orderRepository, EventPublisherRepository eventPublisher) {
        this.orderRepository = orderRepository;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    public Order execute(String orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new NotFoundException("Order not found"));

        try {
            order.confirmOrder();
        } catch (IllegalArgumentException e) {
            throw new BusinessRuleException(e.getMessage());
        }

        orderRepository.update(order);

        try {
            eventPublisher.publishConfirmedOrder(order);
        } catch (Exception e) {
            log.error("Failed to publish OrderConfirmed for order {}", order.getId(), e);
        }

        return order;
    }
}
